### DISCUSSION TITLE
How to use a different resampler on Windows

### DISCUSSION THREAD

**Discussion by sweetblue-ler** [MusicPlayerDaemon/MPD#1746] [Q&A] [answered]
Hello everyone,

For various reasons (the main one being that I use MPD in order to get track infos with a OBS plugin), I'm using MPD on Windows.

The thing is, I really want to use the crossfade functionality, which, of course, can't be used if all tracks output are not resampled.

So I'm using the following config line :

audio_output_format "48000:16:2"

And the crossfade does work perfectly, but the sound quality is really subpar, and it seems the Windows build don't have anything else than the internal resampler.

Now, I must admit, I'm not used to compiling software, and don't really know where to start to get a build of MPD integrating some better resampler.

Could someone help, please?

**Comment by MaxKellermann** [comment]
Figure out how to build either libsamplerate or libsoxr on Windows, and then integrate these libraries in `win32/build.py`.

**Reply by sweetblue-ler** [reply]
Hi, thank you for your reply.

I managed to make a setup to compile mpd for Windows.

Thus said, I have trouble figuring out how to add libsamplerate to build.py

I naively added "libsamplerate" to the thirdparty_libs list, but starting the compilation warns me that "name 'libsamplerate' is not defined".
I'm not sure what is expected, to be honest.

I tried getting the libsamplerate0 pkg from Debian, and also compiled and installed libsamplerate 0.2.2, but the build.py scripts still can't find the library.

**Reply by MaxKellermann** [reply]
The libsamplerate you have is compiled for Linux. It cannot possibly work on Windows. You have to add build instructions for libsamplerate to the MPD build scripts. Just adding the word "libsamplerate" somewhere cannot work.

**Reply by sweetblue-ler** [reply]
My bad, thank you.

I'm trying to figure out how to integrate the Windows libsamplerate compiled library in the MPD build python script.
I guess I have to figure out what "build.libs" references in the python script, but I'm not used to Python, this is what troubles me for now.

**Reply by MaxKellermann** [reply]
No, "Windows libsamplerate compiled library" is wrong. I wrote you need to add build instructiosn for libsamplerate. Not "Windows libsamplerate compiled library".

**Comment by sweetblue-ler** [comment, marked as answer]
I worded my sentence poorly, but I undersood what I needed to do.

I figured it our this morning, after adding "libsamplerate" to the  thirdparty_libs list in the build python script, I added an entry for libsamplerate in the python/build/libs.py script.

I retrieved the latest release of libsamplerate and used Cmake to integrate it to MPD :

```
libsamplerate = CmakeProject(
'https://github.com/libsndfile/libsamplerate/releases/download/0.2.2/libsamplerate-0.2.2.tar.xz',
'97c010fc25156c33cddc272c1935afab',
'lib/libsamplerate.a',
    [
        '-DBUILD_SHARED_LIBS=OFF',

        '-DINSTALL_DOCS=OFF',

        '-DINSTALL_CMAKE_PACKAGE_MODULE=OFF',
    ],
)
```

It worked like a charm, I can finally use MPD with libsamplerate on Windows.

**Reply by MaxKellermann** [reply]
Why don't you submit a pull request?

**Reply by sweetblue-ler** [reply]
I'd love to, but as you may have guessed by now, I'm a complete beginner when it comes to software development =)
I only have surface level knowledge and don't know how to properly share this with a pull request.

But I would really like it if my efforts would benefit other people, yes.

**Reply by sweetblue-ler** [reply]
Allright, I took a bit of time to try and do a pull request, I hope it can help others.
